### PR TITLE
perf: bunch of features and optimizations for top-k (sampling + sparse attention)

### PR DESCRIPTION
**Pull request by yzh119**
## 📌 Description

This PR implements several features and optimizations for top-k:

### Multi-CTA optimization

Followup of #2044 , this PR optimizes the top-k/top-p filter using multi-cta optimizations, more specifically:
* split the vocabulary into chunks and let each cta handles one chunk
* make sure the logits/probs can fit into shared memory inside a CTA
* use global memory to store data structures for cross-cta synchronization.
* make sure the total number of ctas (a multiple of num_chunks) do no exceed number of SMs, using loop to iterate over rows when batch size is greater than number of groups (num_ctas/num_chunks).

The major advantage over the main branch is that we make sure logits/probabilities are stored in shared memory, so multi-round of scan doesn't affect performance too much.

### Radix Top-K

The number of radix bits is set to 8.

### More Top-K APIs and Data Types

* Data types: We used to only support fp32 and this PR implements fp16 and bf16.
* APIs: besides top-k mask logits/renorm probs, we add APIs for raw top-k (return indices/values), and fusion with page table construction (for sparse attention).

## Speedup

On H100, the speedup over 0.5.2 is available at:
https://docs.google.com/spreadsheets/d/1DO8_11gzv-EUACCY6q4IMIHa8SaYv4q8hJ6gZl-D0mU/edit?usp=sharing (3-14x faster).

On consumer GPUs (e.g. Ada6000), the gap is even larger, e.g. for small batch size and large vocabulary setting:
```
v0.5.2
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=1), k: 10, duration: 3004.42 us, effective bandwidth: 2.73 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=1), k: 100, duration: 3633.15 us, effective bandwidth: 2.25 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=1), k: 1000, duration: 4258.82 us, effective bandwidth: 1.92 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=1), k: 5000, duration: 4256.77 us, effective bandwidth: 1.92 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=5), k: 10, duration: 2376.80 us, effective bandwidth: 3.45 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=5), k: 100, duration: 3627.01 us, effective bandwidth: 2.26 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=5), k: 1000, duration: 3945.38 us, effective bandwidth: 2.08 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=5), k: 5000, duration: 4259.84 us, effective bandwidth: 1.92 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=0.1), k: 10, duration: 3316.74 us, effective bandwidth: 2.47 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=0.1), k: 100, duration: 3624.96 us, effective bandwidth: 2.26 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=0.1), k: 1000, duration: 4566.02 us, effective bandwidth: 1.79 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=0.1), k: 5000, duration: 4576.26 us, effective bandwidth: 1.79 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=1), k: 10, duration: 3003.39 us, effective bandwidth: 2.73 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=1), k: 100, duration: 4260.86 us, effective bandwidth: 1.92 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=1), k: 1000, duration: 3947.52 us, effective bandwidth: 2.08 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=1), k: 5000, duration: 5514.24 us, effective bandwidth: 1.49 GB/s

Multi-CTA optimization (only) 
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=1), k: 10, duration: 322.56 us, effective bandwidth: 25.40 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=1), k: 100, duration: 388.10 us, effective bandwidth: 21.11 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=1), k: 1000, duration: 455.68 us, effective bandwidth: 17.98 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=1), k: 5000, duration: 455.68 us, effective bandwidth: 17.98 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=5), k: 10, duration: 257.02 us, effective bandwidth: 31.87 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=5), k: 100, duration: 388.10 us, effective bandwidth: 21.11 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=5), k: 1000, duration: 421.89 us, effective bandwidth: 19.42 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=5), k: 5000, duration: 455.68 us, effective bandwidth: 17.98 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=0.1), k: 10, duration: 355.33 us, effective bandwidth: 23.05 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=0.1), k: 100, duration: 388.10 us, effective bandwidth: 21.11 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=0.1), k: 1000, duration: 486.40 us, effective bandwidth: 16.84 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=0.1), k: 5000, duration: 488.45 us, effective bandwidth: 16.77 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=1), k: 10, duration: 320.51 us, effective bandwidth: 25.56 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=1), k: 100, duration: 452.61 us, effective bandwidth: 18.10 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=1), k: 1000, duration: 422.91 us, effective bandwidth: 19.37 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=1), k: 5000, duration: 585.73 us, effective bandwidth: 13.99 GB/s

Multi-CTA + Radix (this PR)
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=1), k: 10, duration: 29.70 us, effective bandwidth: 275.86 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=1), k: 100, duration: 28.67 us, effective bandwidth: 285.71 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=1), k: 1000, duration: 29.70 us, effective bandwidth: 275.86 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=1), k: 5000, duration: 28.67 us, effective bandwidth: 285.71 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=5), k: 10, duration: 29.70 us, effective bandwidth: 275.86 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=5), k: 100, duration: 28.67 us, effective bandwidth: 285.71 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=5), k: 1000, duration: 29.70 us, effective bandwidth: 275.86 GB/s
vocab_size: 256000, batch_size: 4, distrib: normal_distribution(std=5), k: 5000, duration: 28.67 us, effective bandwidth: 285.71 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=0.1), k: 10, duration: 28.67 us, effective bandwidth: 285.71 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=0.1), k: 100, duration: 28.90 us, effective bandwidth: 283.50 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=0.1), k: 1000, duration: 28.67 us, effective bandwidth: 285.71 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=0.1), k: 5000, duration: 29.66 us, effective bandwidth: 276.16 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=1), k: 10, duration: 28.67 us, effective bandwidth: 285.71 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=1), k: 100, duration: 28.67 us, effective bandwidth: 285.71 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=1), k: 1000, duration: 28.67 us, effective bandwidth: 285.71 GB/s
vocab_size: 256000, batch_size: 4, distrib: gumbel_distribution(beta=1), k: 5000, duration: 28.86 us, effective bandwidth: 283.81 GB/s
```
The gap can be as large as 100 times.

## 🔍 Related Issues

#2044 

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * Added a radix-based GPU top-k API and a public top_k function (returns values and indices, optional sorting).

* **Improvements**
  * New multi-CTA top-k path with per-call row-states buffer for better large-batch throughput; FP32/FP16/BF16 supported.
  * Sampling ops and logits/probs top-k now accept and allocate a row-states buffer; small utility helpers and NVCC flags updated.
  * Exposed topk JIT module and top_k at package level.

* **Tests**
  * New and expanded tests for top-k and sampling across dtypes, large batches, and relaxed numeric tolerances.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: include/flashinfer/sampling.cuh [hunk 1]
```diff
@@ -2071,6 +2071,446 @@ cudaError_t TopKMaskLogits(DType* logits, DType* masked_logits, IdType* top_k_ar
...
+  asm volatile("ld.cg.global.b32 %0, [%1];\n" : "=r"(state) : "l"(ptr));
+#endif
+
+  return state;
+}
+
+__device__ __forceinline__ void red_release(int* ptr, int val) {
+#if (__CUDA_ARCH__ >= 700)
+  // SM70 and newer use memory consistency qualifiers
+  // Release pattern using acq_rel fence + relaxed modifier
```
Comment (nvmbreughe) [inline include/flashinfer/sampling.cuh:1843]: Maybe add some more clarifiction: besides releasing, this also performs a reduction (sum)

File: include/flashinfer/sampling.cuh [hunk 2]
```diff
@@ -2071,6 +2071,446 @@ cudaError_t TopKMaskLogits(DType* logits, DType* masked_logits, IdType* top_k_ar
...
+  size_t logits_offset = ((temp_storage_size + 15) / 16) * 16;
+  DType* shared_logits = reinterpret_cast<DType*>(smem + logits_offset);
+
+  // Note: arrival_counter and count buffers should be pre-initialized to zero on the host side
+
+  // Persistent iteration counter for double buffering (never resets across rows)
+  int persistent_iteration = 0;
+
+  // Calculate total number of iterations for persistent loop
+  uint32_t num_groups = gridDim.x / ctas_per_group;
```
Comment (nvmbreughe) [inline include/flashinfer/sampling.cuh]: Shoud we add tests that explicitly trigger num_groups > 1 ?

Comment (yzh119) [reply include/flashinfer/sampling.cuh]: we do have, for vocab_size=128256, it will be splitted into 4 chunks (one per SM), so when batch_size is greater than 33, the num_groups will be greater than 132 (the number of SMs on hopper) for H100.
